feat: tag sections, subsections, and the whole course (TEMP)

rpenido opened the pull request.

PKulkoRaccoonGang committed:
feat: [AXIMST-23] Course unit - Sidebar with unit info (#117)

* feat: added Sidebar with unit info

* feat: added unit location

* refactor: added legacy behavior

* feat: added live variant

* refactor: code refactoring

* feat: added tests and translations

* feat: added new font size

* refactor: after review

ihor-romaniuk committed:
feat: [FC-0044] Unit page - display xblock components

PKulkoRaccoonGang committed:
feat: [AXIMST-25] Course unit - Alert notification about unpublished changes (#135)

* feat: [AXIMST-25] added alert notification about unpublished changes

* feat: added tests

* feat: added translations

PKulkoRaccoonGang committed:
feat: [AXIMST-24] Course unit - Sidebar buttons functional (#134)

* feat: [AXIMST-24] sidebar buttons functional

* refactor: removed modal extra className

* refactor: refactoring after review

PKulkoRaccoonGang committed:
fix: [AXIMST-427] fixed unpublished changes alert

PKulkoRaccoonGang committed:
fix: [AXIMST-470] fixed sidebar status after deleting or duplicating xblock

PKulkoRaccoonGang committed:
fix: [AXIMST-472] fixed sidebar visibility notification

PKulkoRaccoonGang committed:
fix: [AXIMST-473] fixed sidebar publish status

PKulkoRaccoonGang committed:
refactor: after rebase

PKulkoRaccoonGang committed:
refactor: refactoring after review

rpenido commented on src/course-outline/CourseOutline.jsx:
```diff
@@ -158,26 +159,19 @@ const CourseOutline = ({ courseId }) => {
-    data: unitsTagCounts,
-    isSuccess: isUnitsTagCountsLoaded,
-  } = useUnitTagsCount(unitsIdPattern);
+  // ToDo: Discuss in the PR
+  // We could use this to prefetch the tags count for all units/blocks in the course
+  // But to prevent refetching the same data when the component is re-rendered, we should
+  // configure the query stale time to a reasonable value.
+  // IMHO, we should just let each card to fetch the tags count by itself. This is simpler and
+  // is working fine.
+  // const contentTagPatternMutation = useContentPatternTagsCount();
+  //
+  // useEffect(() => {
+  //   if (courseId) {
+  //     contentTagPatternMutation.mutate(`${courseId.replace('course-v1', 'block-v1', 1)}*`);
+  //   }
+  // }, [courseId]);
```
@ChrisChV I think we don't need this. The multiple calls were a problem in the Studio, but I think it works smoothly here. What do you think? CC @bradenmacdonald 

PS: This is not fully working right now, but I think I can fix it.

ChrisChV replied:
@rpenido I think we don't need it for now, it's a lot for an MVP. Unless a performance analysis says flatly otherwise :smile: 

About let each cart to fetch the tag count, I think it would increase the number of calls made to the API, one for each card. Using one or the other approach I don't see much difference, but the increase in the number of calls is what makes me lean toward just making one call for all

CC @bradenmacdonald

rpenido replied:
I think it is working smoothly even when making A LOT of API calls.

[tag_course_outline.webm](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/629c4802-2191-4ad9-a7b1-e1c52406502a)

We could tweak it by setting a stale time for each query, but I think we should do this for the whole app instead of on a query-by-query basis.

bradenmacdonald replied:
Our API to get the counts for every item in the course should be pretty fast. I would lean toward just using that everywhere, so there's only one API request. Fetching hundreds of separate API requests on the outline page may work fine on devstack but I suspect will cause issues on production. Especially if people don't have HTTP/2 working.

The only downside is when you invalidate the tags of a particular item, you also have to invalidate the counts of everything in the whole course and fetch it again, but I think it's fine.

Also to be clear, each component (e.g. each unit in the outline) should use the same API hook to "fetch tag counts for the whole course". React Query will de-duplicate this to only a single request. Then take the result, search for the ID of the current unit, and display that count, if any.

rpenido replied:
> The only downside is when you invalidate the tags of a particular item, you also have to invalidate the counts of everything in the whole course and fetch it again, but I think it's fine.

It is possible to solve this: A mutation prefetch the data and populate the queries via `setQueryData`. So we get all counts on first load and then each component have their own query after that.

https://github.com/open-craft/frontend-app-course-authoring/pull/30/files#diff-0fb42c9f642840f13f2018e2ab11fc2386e22caa9baa226843197bf16d4151cfR10-R24

The problem with this code is that mounting the component (the unit cards when expanding) trigger the refetch for the query. This could be fixed setting a proper `staleTime`, but it will need more time.

I will leave it for now and use a single useQuery as you both suggested (but we could go back to this mixed approach if someday the single query performance hit us).

Thank you for the feedback @ChrisChV @bradenmacdonald !

bradenmacdonald replied:
@rpenido I thought about that too, and it's what I wanted to suggest, but I think there's a problem with it. During the pre-fetch, if the individual cards render before the prefetch is complete, they will still each individually make a request to the API. I looked a lot of the React Query docs like https://tanstack.com/query/v4/docs/framework/react/guides/prefetching and https://tkdodo.eu/blog/seeding-the-query-cache but I couldn't figure out a way to tell React Query _not_ to load the per-item query keys while the prefetch is happening, but still to load them later if they get invalidated.

rpenido replied:
> but I couldn't figure out a way to tell React Query _not_ to load the per-item query keys while the prefetch is happening, but still to load them later if they get invalidated.

We could call the prefetch mutation at the CourseOutline page and render (the page or the course tree) only after we get the mutation response, as we do with all course data (sections, subsections, etc.).

I think it will be possible if we spend some time with it, but I don't think we should do it now. The current overall responsiveness is pretty good. I don't want to add complexity to solve a problem we don't have yet. :slightly_smiling_face:

rpenido commented on src/course-outline/card-header/CardHeader.jsx:
```diff
@@ -91,134 +97,148 @@ const CardHeader = ({
+        </div>
       </div>
-    </div>
+      <Sheet
+        position="right"
+        show={isManageTagsDrawerOpen}
+        onClose={/* istanbul ignore next */ () => closeManageTagsDrawer()}
+      >
+        <ContentTagsDrawer
+          id={cardId}
+          onClose={/* istanbul ignore next */ () => closeManageTagsDrawer()}
+        />
+      </Sheet>
+    </>
   );
```
I moved the Drawer, the tagCount and the related states and events to the `CardHeader`.

Let me know what you think!

rpenido commented on src/course-outline/card-header/CardHeader.jsx:
```diff
@@ -50,16 +54,18 @@ const CardHeader = ({
   const isDisabledPublish = (status === ITEM_BADGE_STATUS.live
     || status === ITEM_BADGE_STATUS.publishedNotLive) && !hasChanges;
 
+  const { data: contentTagCount, isSuccess: isContentTagCountSuccess } = useContentTagsCount(cardId);
+
```
Each item asks for its tag count. As commented above, we could prefetch it (to prevent multiple API calls), but I think we don't need it.

KristinAoki committed:
feat: add checklist page (#870)

* feat: add checklist page

* fix: failing tests

* fix: styling bugs

* fix: lint errors

* feat: add test fro CourseChecklist

* fix: lint errors

* feat: add ChecklistSection tests

* fix: lint error

* fix: missing api reply status

jesperhodge committed:
Fix  radio context provider missing (#874)

This is a temporary fix for a bug that stems from a paragon component, SelectableBox. So we're using our own copy from fronten-lib-content-components.

KristinAoki committed:
fix: accessibility page config reference (#875)

KristinAoki committed:
fix: revert addition of course-ingestion-tool submodule (#878)

ChrisChV committed:
[UU-58] Implement tagging & taxonomy feature in outline (#855)

* feat: TagCount component

* feat: Update ContentTagsDrawer to use it in the MFE

* feat: Manage tags menu added on units

* feat: Tag count added on unit

* feat: Add button feat to Tag count

* test: Course Outline api tests

* test: Ignore lines that can not be tested

* style: Comment added on ContentTagsDrawer

* style: Nits on CardHeader

rpenido committed:
feat: tag sections, subsections, and the whole course

rpenido commented:
Closed if favour of:
- https://github.com/openedx/frontend-app-course-authoring/pull/879
